src/expr/src/interpret.rs

```diff
@@ -898,41 +898,74 @@ impl<'a> Interpreter for ColumnSpecs<'a> {
 /// An interpreter that returns whether or not a particular expression is "pushdownable".
 /// Broadly speaking, an expression is pushdownable if the result of evaluating the expression
 /// depends on the range of possible column values in a way that `ColumnSpecs` is able to reason about.
-/// `Column` is trivially pushdownable; `Literal` and `CallUnmaterializable`
-/// are not, since their results don't depend on the range of possible values in any column.
 ///
-/// Functions are the most complex case. We say that a function is pushdownable for a particular
-/// argument if `ColumnSpecs` can determine the spec of the function's output given the input spec for
-/// that argument. (In practice, this is true when either the function is monotone in that argument
-/// or it's been special-cased in the interpreter.) And we say a function-call expression is
-/// pushdownable overall if at least one of its arguments is pushdownable, and the function is
-/// is pushdownable in that argument. For example, `3 + #0` is pushdownable (because the second
-/// argument to `+` is pushdownable and `+` is monotone) but `3 / #0` is not (because while the
-/// second argument is pushdownable, `/` is not monotone on the right-hand side).
+/// In practice, we internally need to distinguish between expressions that are trivially predicable
+/// (because they're constant) and expressions that depend on the column ranges themselves.
+/// See the [TraceSummary] variants for those distinctions, and [TraceSummary::pushdownable] for
+/// the overall assessment.
 #[derive(Debug)]
 pub struct Trace;
 
+/// A summary type for the [Trace] interpreter.
+// NB: the ordering here is important!
+#[derive(Copy, Clone, Debug, PartialOrd, PartialEq, Ord, Eq)]
+pub enum TraceSummary {
+    /// The expression is constant: we can evaluate it without any runtime information.
+    Constant,
+    /// The expression depends on runtime information, but in "predictable" way... ie. if we know
+    /// the range of possible values for all columns and unmaterializable functions, we can
+    /// predict the possible values of the output.
+    Dynamic,
+    /// The expression depends on runtime information in an unpredictable way.
```

Comment:
nit: expand what you mean by "an unpredictable way". IIUC this entails that knowledge about value ranges of the inputs cannot be used to infer the value range of the output.